### PR TITLE
Make ContextKey first-class

### PR DESCRIPTION
**Pull request by bricelam**
Part of #2635

### CODE REVIEW

File: src/EntityFramework7.Core/Internal/ModelSource.cs [hunk 1]
```diff
@@ -46,7 +46,8 @@ public virtual IModel GetModel(DbContext context, IConventionSetBuilder conventi
 
             var productVersion = typeof(ModelSource).GetTypeInfo().Assembly
                 .GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
-            model[CoreAnnotationNames.ProductVersionAnnotation] = productVersion;
+            model.SetProductVersion(productVersion);
+            model.SetContextKey(context.GetType().FullName);
```
Comment (smitpatel) [inline src/EntityFramework7.Core/Internal/ModelSource.cs]: Should use `DisplayName()` function from `TypeExtensions` for better handling of nested classes.

Comment (bricelam) [inline src/EntityFramework7.Core/Internal/ModelSource.cs]: Then I can't do crazy hacks like this inside my provider. :wink: 
```C#
var context = Activator.CreateInstance(Type.GetType(model.GetContextKey()));
```
But on a serious note, is it worth it? It seems to only make generic types prettier and you can't use open generic types with Migrations anyway.

Comment (divega) [inline src/EntityFramework7.Core/Internal/ModelSource.cs]: So it doesn't not help with nested types?

Comment (smitpatel) [inline src/EntityFramework7.Core/Internal/ModelSource.cs]: As long as the string `Type.FullName` is kept intact, there won't be any issue.
Considering above usage, it is not worth. :smile:

Comment (ajcvickers) [inline src/EntityFramework7.Core/Internal/ModelSource.cs]: I didn't think we still had a restriction that the DbContext could not be a generic type? Or a nested type in a generic type for that matter. Is it only Migrations that imposes this restriction?

Comment (ajcvickers) [inline src/EntityFramework7.Core/Internal/ModelSource.cs]: I found a test that does `using (var context = new BloggingContext<TBlog>()` so it is at least nominally supported.

Comment (bricelam) [inline src/EntityFramework7.Core/Internal/ModelSource.cs]: We need to work out how to get an instance of it, but after that, you're right, it should be possible.

Comment (bricelam) [inline src/EntityFramework7.Core/Internal/ModelSource.cs]: Will update to use `DisplayName` and make a note to test generic context types with Migrations.

Comment (bricelam) [inline src/EntityFramework7.Core/Internal/ModelSource.cs]: @divega No, just generic types.

**Comment by bricelam** [comment]
Revoking. We're going to try getting rid of the ContextKey instead.

**Pull request closed without merging**